### PR TITLE
Cherry-pick of #101154 - Adding IPV6 (Dual Stack) support to handle IPV6 pod cidrs

### PR DESCRIPTION
**Pull request by basantsa1989**
Adding IPV6 (Dual Stack) support to handle both IPV4 and IPV6 pod cidrs in cloud_cidr_allocator

Cherry-pick of https://github.com/kubernetes/kubernetes/pull/101154

/kind bug

```release-note
NONE
```

### CODE REVIEW

**Comment by bowei** [comment]
/lgtm

work with the release team to complete cherry pick

**Comment by justaugustus** [comment]
> @kubernetes/release-managers please let me know the next steps.

@basantsa1989 -- Needs an `/approve` label from the area's OWNERS before this can be reviewed by Release Managers for cherry pick approval.

See https://github.com/kubernetes/kubernetes/pull/102574#issuecomment-854168230 and https://github.com/kubernetes/kubernetes/pull/102574#issuecomment-854169985 for more details.

/assign @bowei @cheftako

File: pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go [hunk 1]
```diff
@@ -246,50 +248,56 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
...
 	if node.Spec.ProviderID == "" {
 		return fmt.Errorf("node %s doesn't have providerID", nodeName)
 	}
 
-	cidrs, err := ca.cloud.AliasRangesByProviderID(node.Spec.ProviderID)
+	cidrStrings, err := ca.cloud.AliasRangesByProviderID(node.Spec.ProviderID)
 	if err != nil {
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
-		return fmt.Errorf("failed to allocate cidr: %v", err)
+		return fmt.Errorf("failed to get cidr(s) from provider: %v", err)
```
Comment (cheftako) [inline pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go:261]: If we're switching to structured logging, shouldn't this switch as well?

File: pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go [hunk 2]
```diff
@@ -246,50 +248,56 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
...
-	if len(cidrs) == 0 {
+	if len(cidrStrings) == 0 {
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: Node %v has no CIDRs", node.Name)
 	}
-	_, cidr, err := net.ParseCIDR(cidrs[0])
+	//Can have at most 2 ips (one for v4 and one for v6)
+	if len(cidrStrings) > 2 {
+		klog.InfoS("Got more than 2 ips, truncating to 2", "cidrStrings", cidrStrings)
+		cidrStrings = cidrStrings[:2]
```
Comment (cheftako) [inline pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go:270]: If we have more than 2 ips, how do we know what the first two entries would be? We might get two ipv4s and then an ipv6.

File: pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go [hunk 3]
```diff
@@ -246,50 +248,56 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
...
+	}
+
+	cidrs, err := netutils.ParseCIDRs(cidrStrings)
 	if err != nil {
-		return fmt.Errorf("failed to parse string '%s' as a CIDR: %v", cidrs[0], err)
+		return fmt.Errorf("failed to parse strings %v as CIDRs: %v", cidrStrings, err)
```
Comment (cheftako) [inline pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go:275]: structured?

**Comment by basantsa1989** [comment]
@cheftako thank you for the review. 
Sorry, I had sent this PR out way back assuming legacy-cloud-provider was open for commits.

I have this equivalent change in 1.22 onwards and we will not use 1.21 for IPv6. 
Let me know if you see any other reasons we may need this in 1.21. 

I'll close this PR. I'll take your comments and address them on the master.

Thanks!

**Pull request closed without merging**